### PR TITLE
Improve: Make serde optional, a dirty work

### PR DESCRIPTION
**Pull request by PsiACE**
Signed-off-by: Chojan Shang <psiace@outlook.com>

This is a simple implementation and serde will be the default feature. But of course, you can turn it off now with 'default-features = false'.

I need some suggestions to see how to make it better.

**Checklist**
- [ ] Updated guide with pertinent info (may not always apply). 
- [x] Squash down commits to one or two logical commits which clearly describe the work you've done. If you don't, then Dodd will `:)`.

---
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/243)

### CODE REVIEW

File: openraft/src/lib.rs [hunk 1]
```diff
@@ -85,9 +83,15 @@ pub use crate::vote::Vote;
 /// ## Note
 ///
 /// The trait is automatically implemented for all types which satisfy its supertraits.
-pub trait AppData: Clone + Send + Sync + Serialize + DeserializeOwned + 'static {}
+#[cfg(feature = "serde")]
+pub trait AppData: Clone + Send + Sync + serde::Serialize + serde::de::DeserializeOwned + 'static {}
+#[cfg(not(feature = "serde"))]
+pub trait AppData: Clone + Send + Sync + 'static {}
```
Comment (PsiACE) [inline openraft/src/lib.rs:89]: A part of the code has been rewritten, it just works, but not elegantly enough.

**Review by schreter — COMMENTED** [review]
LGTM

> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/243#-:-MxYs0J6BG9xUZPJkyqP:bb6ywmx)* status: all files reviewed, 1 unresolved discussion (waiting on @PsiACE)

___
*[openraft/src/lib.rs, line 89 at r1](https://reviewable.io/reviews/datafuselabs/openraft/243#-MxYpN0FGkOiZNUu3rYj-r1-89:-MxYqISP7FnUf303XjWt:bv8f3je) ([raw file](https://github.com/datafuselabs/openraft/blob/f62f19ba5fe921bbd7eeb2cd1418f9a39e6d1256/openraft/src/lib.rs#L89)):*
<details><summary><i>Previously, PsiACE wrote…</i></summary><blockquote>

A part of the code has been rewritten, it just works, but not elegantly enough.
</blockquote></details>

Well, what you could do here is to define a supertrait `RaftSerializable`, which would be empty for non-serde implementation and implemented for any `T` and use `Serialize + Deserialize` supertraits for serde implementation and implemented for any `T: Serialize + Deserialize`. Then, make it supertrait of `AppData` and `AppDataResponse` and other types, as needed. It will reduce the amount of changes somewhat should the requirements change or another serialization should be supported.

But I think the current approach is good enough :-).
___
*[openraft/src/metrics.rs, line 31 at r2](https://reviewable.io/reviews/datafuselabs/openraft/243#-MxYrf4W-UrJ-jj-KTXK:-MxYrf4W-UrJ-jj-KTXL:bkkzpyy) ([raw file](https://github.com/datafuselabs/openraft/blob/a9dc10c24e0464590dfe87818a41b6a02fcf112b/openraft/src/metrics.rs#L31)):*
> ```XML
> /// A set of metrics describing the current state of a Raft node.
> #[derive(Clone, Debug, PartialEq, Eq)]
> #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
> ```

The derives could be possibly simplified with an own derive macro so you don't have to type so much and also to possibly support other serialization frameworks in the future by changing it in one place, but I think that's an overkill for the first version and this is good enough.

**Review by schreter — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/243)* status: all files reviewed, 2 unresolved discussions (waiting on @PsiACE)

___
*[openraft/Cargo.toml, line 45 at r2](https://reviewable.io/reviews/datafuselabs/openraft/243#-MxYsBoZ3XZ3-gLUBxg8:-MxYsBoZ3XZ3-gLUBxg9:barqs8k) ([raw file](https://github.com/datafuselabs/openraft/blob/a9dc10c24e0464590dfe87818a41b6a02fcf112b/openraft/Cargo.toml#L45)):*
> ```TOML
> 
> [features]
> default = ["serde"]
> ```

What is missing, though, is a test crate which uses openraft without the default feature to verify that it won't get broken.

Not sure what is the best approach here, since so far there are no extra test crates yet. Probably add a binary test crate `test_no_serde` or so in the workspace, which won't get published to crates.io, just used in the build of the workspace to ensure compatibility.

NOTE: just adding a test won't help, since the features are cumulative and cargo would activate serde for the test. You indeed need a crate which doesn't have other dependencies and just uses openraft w/o default features. Then you can test that the types don't implement `serde::Serialize` by using something which doesn't use `serde` as `AppData`, for instance. You don't need to spin up Raft, just do the compile check there (i.e., one-liner `main()` which tries to convert a reference to some non-serde type to &dyn AppData).

File: openraft/Cargo.toml [hunk 2]
```diff
@@ -42,6 +42,8 @@ tracing-subscriber = { version = "0.3.3",  features=["env-filter"] }
 
 
 [features]
+default = ["serde"]
+
```
Comment (PsiACE) [inline openraft/Cargo.toml:46]: hi @schreter , thank you for the kind review. I think we can disable `serde` by default. Then use [serde_test](https://crates.io/crates/serde_test) to do some tests.

What do you think?

Comment (schreter) [reply openraft/Cargo.toml:46]: > I think we can disable serde by default. Then use [serde_test](https://crates.io/crates/serde_test) to do some tests.

Of course, *not* using `serde` by default and only activating it if the `serde` feature is on is the other (possibly preferable) option, but I'm not sure what is the best. That should be decided by one of the maintainers.

Independent of that, you still need to create an extra crate to test it with the feature enabled, then. It is to my knowledge simply impossible to test both with a feature enabled and a feature disabled inside of the same crate.

@drmingdrmer Opinions/directions?

Comment (drmingdrmer) [reply openraft/Cargo.toml:46]: @schreter @PsiACE 
Disabling `serde`  by default looks good to me.

> Independent of that, you still need to create an extra crate to test it with the feature enabled, then. It is to my knowledge simply impossible to test both with a feature enabled and a feature disabled inside of the same crate.

I'm used to writing feature flag related unit tests like this:
```rust
#[cfg(feature = "backtrace")]
#[test]
fn test_backtrace() -> anyhow::Result<()> {
    let got = backtrace_str().expect("no none");
    assert!(got.contains("test_backtrace"));
    Ok(())
}

#[cfg(not(feature = "backtrace"))]
#[test]
fn test_backtrace() -> anyhow::Result<()> {
    let got = backtrace_str();
    assert!(got.is_none());
    Ok(())
}
```

And let CI run it twice: with such a feature flag on and off. 

Good enough?

Comment (schreter) [reply openraft/Cargo.toml:46]: > Good enough?

Sure, it just then takes double resources in the CI :-). But if there are sufficient resources, then this indeed is the best approach.

Comment (PsiACE) [reply openraft/Cargo.toml:46]: > Good enough?

I've written this a lot in the past, and I think it's totally feasible.

**Comment by PsiACE** [comment]
A feature called `"serde_impl"` is currently used to control serde support. I haven't added any relevant tests for now, because I'm not sure which ones are necessary.

Since `memstore` and `example-raft-kv` both rely on `openraft/serde_impl`, I think they have helped us with a simple verification. Please remind me if I've missed anything.

File: openraft/src/raft_types.rs [hunk 3]
```diff
@@ -138,7 +137,8 @@ impl Display for SnapshotSegmentId {
 }
 
 // An update action with option to update with some value or just leave it as is.
-#[derive(Debug, Clone, PartialOrd, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialOrd, PartialEq, Eq)]
+#[cfg_attr(feature = "serde_impl", derive(serde::Deserialize, serde::Serialize))]
 pub enum Update<T> {
     Update(T),
     AsIs,
```
Comment (drmingdrmer) [inline openraft/src/raft_types.rs:144]: `Update<T>` does not need to be serializable.
It must have been added by accident. :(

**Review by drmingdrmer — COMMENTED** [review]
`memstore` and `example-raft-kv` may not covers every types.
We need to ensure at least these types to be serializable(top-level types) in unit test(`openraft` unitest should be run twice by github workflow CI: with and without feature `serde-impl`):

- Essential types:
`Entry`
`RaftMetrics`
`Node`

- Trait implementations:
SomeType that impl `NodeId`
SomeType that impl `AppData`
SomeType that impl `AppDataResponse`

- User API errors:
`InitializeError`
`AddLearnerError`
`ChangeMembershipError`
`ClientWriteError`
`CheckIsLeaderError`
`Fatal`

- Types used by Raft protocol API:
`RPCError`.
Request, Response and Errors that are used in `RaftNetwork`.

**Comment by drmingdrmer** [comment]
> Edit:
> The types used in `Raft` should also be included:
> `AddLearnerResponse`
> `ClientWriteResponse`

Add two missing serializble types

**Pull request closed without merging**